src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -433,6 +434,90 @@ impl LogicalJoin {
         )
     }
 
+    /// find the comparison exprs and return their inputs, types and indices.
+    fn find_comparison_exprs(
+        left_col_num: usize,
+        right_col_num: usize,
+        exprs: &[ExprImpl],
+    ) -> (Vec<ExprImpl>, Vec<ExprImpl>, Vec<(usize, Type)>) {
+        let left_bit_map = FixedBitSet::from_iter(0..left_col_num);
+        let right_bit_map = FixedBitSet::from_iter(left_col_num..left_col_num + right_col_num);
+
+        let mut left_exprs = vec![];
+        let mut right_exprs = vec![];
+        // indices and return types of function calls whose's inputs will be calculated in
+        // `project`s
+        let mut indices_and_ty_of_func_calls = vec![];
+        let is_comparison_type = |ty| {
+            matches!(
+                ty,
+                Type::LessThan
+                    | Type::LessThanOrEqual
+                    | Type::Equal
+                    | Type::GreaterThan
+                    | Type::GreaterThanOrEqual
+            )
+        };
+        for (index, expr) in exprs.iter().enumerate() {
+            if let ExprImpl::FunctionCall(func) = expr && is_comparison_type(func.get_expr_type()) {
+                let (ty, left, right) = func.clone().decompose_as_binary();
+                // we just cast the return types of inputs of equal conditions for `HashJoin`.
+                // non-equal conditions don't need unnecssary explicit cast.
+                let can_cast = ty == Type::Equal;
+                let left_input_bits =
+                    left.collect_input_refs(left_col_num + right_col_num);
+                let right_input_bits =
+                    right.collect_input_refs(left_col_num + right_col_num);
+                let (mut left, mut right) = if left_input_bits.is_subset(&left_bit_map)
+                    && right_input_bits.is_subset(&right_bit_map)
+                {
+                    (left, right)
+                } else if left_input_bits.is_subset(&right_bit_map)
+                    && right_input_bits.is_subset(&left_bit_map)
+                {
+                    (right, left)
+                } else {
+                    continue;
+                };
+                // when both `left` and `right` are `input_ref`, and they have the same return type or `func` is non-equal comparison expression, there is no need to calculate them in project.
+                if left.as_input_ref().is_some() && right.as_input_ref().is_some() && (left.return_type() == right.return_type() || !can_cast) {
+                    continue;
+                }
+                if can_cast {
+                    // align return types to avoid error when executing join.
+                    let mut v = vec![left, right];
+                    let _data_type = align_types(v.iter_mut()).unwrap();
```

Comment:
```suggestion
                    align_types(v.iter_mut()).unwrap();
```

Reply:
A simpler way:
```
align_types([&mut left, &mut right].into_iter()).unwrap();
```
This avoids intermediate `v` and `next_tuple` as well.